staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/unstructured_conversion_test.go

```diff
@@ -0,0 +1,465 @@
+			targetGV: schema.GroupVersion{Group: "", Version: "v1"},
+			expectedConvertedObject: &testapigroupv1.Carp{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "v1",
+					Kind:       "Carp",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "noxu",
+				},
+				Spec: testapigroupv1.CarpSpec{
+					Hostname: "example.com",
+				},
+			},
+		},
+		{
+			name: "convert valid versioned unstructured to internal should succeed but ignore type meta",
+			unstructuredToConvert: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Carp",
+					"metadata": map[string]interface{}{
+						"creationTimestamp": nil,
+						"name":              "noxu",
+					},
+					"spec": map[string]interface{}{
+						"hostname": "example.com",
+					},
+					"status": map[string]interface{}{},
+				},
+			},
+			targetGV: schema.GroupVersion{Group: "", Version: "__internal"},
+			expectedConvertedObject: &testapigroup.Carp{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "noxu",
+				},
+				Spec: testapigroup.CarpSpec{
+					Hostname: "example.com",
+				},
+			},
+		},
+		{
+			name: "convert valid unstructured to unexisting version should fail",
+			unstructuredToConvert: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Carp",
+					"metadata": map[string]interface{}{
+						"creationTimestamp": nil,
+						"name":              "noxu",
+					},
+					"spec": map[string]interface{}{
+						"hostname": "example.com",
+					},
+					"status": map[string]interface{}{},
+				},
+			},
+			targetGV: schema.GroupVersion{Group: "", Version: "v9"},
+			expectedErrFunc: func(err error) bool {
+				return reflect.DeepEqual(err, runtime.NewNotRegisteredGVKErrForTarget(
+					scheme.Name(),
+					schema.GroupVersionKind{Group: "", Version: "v9", Kind: "Carp"},
+					nil,
+				))
+			},
+			expectedConvertedObject: nil,
+		},
+	}
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			defer func() {
+				v := recover()
+				assert.Equal(t, testCase.expectPanic, v != nil, "unexpected panic")
+			}()
+			// Convert by specifying destination object
+			outObject, err := scheme.ConvertToVersion(testCase.unstructuredToConvert, testCase.targetGV)
+			if testCase.expectedErrFunc != nil {
+				if !testCase.expectedErrFunc(err) {
+					t.Errorf("error mismatched: %v", err)
+				}
+			}
+			assert.Equal(t, testCase.expectedConvertedObject, outObject)
+		})
+	}
+}
+
+func TestUnstructuredToUnstructuredConversion(t *testing.T) {
+	// eventually, we don't want any inter-unstructured conversion happen, but for now, the conversion
+	// just copy/pastes
+	scheme, _ := test.TestScheme()
+	inUnstructured := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "v1",
+			"kind":       "Carp",
+		},
+	}
+	outUnstructured := &unstructured.Unstructured{}
+	err := scheme.Convert(inUnstructured, outUnstructured, nil)
```

Comment:
can we also add cases for passing a target version in?
* external target version that matches the input should work
* external target version that does not match the input should not work?
* internal target version that does not match the input should not work?

Reply:
done